### ISSUE TITLE
Perform sanity checks and follow good coding practices

### ISSUE THREAD

**Issue by evancofer** [Benjamin-Lee/deep-rules#52]
**Have you checked the [list of proposed rules](https://github.com/Benjamin-Lee/deep-rules/issues?q=is%3Aissue+is%3Aopen+label%3Arule) to see if the rule has already been proposed?**

- [x] Yes

**Did you add yourself as a [contributor](https://github.com/Benjamin-Lee/deep-rules/blob/master/contributors.md) by making a pull request if this is your first contribution?**

- [x] Yes, I added myself or am already a contributor

**Feel free to elaborate, rant, and/or ramble.**
When coding DL models, it is important to maintain good software engineering practices. All code should be documented and include rigorous tests. Sanity checks are also useful. For instance, something is probably wrong (e.g. bug in code, ill posed problem, bad hyperparameters) if model training loss does not decrease (i.e. not overfitting) when considering a very small subset of the training data.

**Any citations for the rule?** (peer-reviewed literature preferred but not required)
- https://arxiv.org/pdf/1206.5533v2.pdf
- http://cs231n.github.io/neural-networks-3/#sanitycheck

**Comment by evancofer** [comment]
This slightly subsumes #49

**Comment by rasbt** [comment]
I think it would be good to have your suggestion as s a separate rule, but it is also somewhat connected #42, the fact that we need to usually have a larger/more extensive model selection part when using deep learning as opposed to "traditional" machine learning. In addition, we need to "more babysit" the different model fitting procedures and evaluating the internal procedure ("does it converge?") vs just the external metrics ("what is the prediction accuracy?")

**Comment by evancofer** [comment]
Indeed. In many cases however, sanity checks need to occur before hyperparameter optimization occurs.

**Comment by Benjamin-Lee** [comment]
>  All code should be documented and include rigorous tests.

With respect to this, I just published [a paper](doi.org/10.1371/journal.pcbi.1006561) on the exact topic that might be worth citing. I'm happy to expand further on documentation best practices for DL.

**Comment by agitter** [comment]
I also recommend [*Top considerations for creating bioinformatics software documentation*](https://doi.org/10.1093/bib/bbw134) for software documentation

**Comment by pstew** [comment]
@Benjamin-Lee Congrats! I think this and the paper linked by @agitter would be great to cite here.

**Comment by fmaguire** [comment]
Sort of mentioned in tip 3, need to add this reference to it.
